var/spack/repos/builtin/packages/py-numpy/package.py

```diff
@@ -87,31 +85,39 @@ class PyNumpy(PythonPackage):
     version("1.14.6", sha256="1250edf6f6c43e1d7823f0967416bc18258bb271dc536298eb0ea00a9e45b80a")
     version("1.14.5", sha256="a4a433b3a264dbc9aa9c7c241e87c0358a503ea6394f8737df1683c7c9a102ac")
 
-    variant("blas", default=True, description="Build with BLAS support")
-    variant("lapack", default=True, description="Build with LAPACK support")
-
-    # Based on wheel availability on PyPI
-    depends_on("python@3.9:3.11", when="@1.25:", type=("build", "link", "run"))
+    depends_on("python@3.9:3.12", when="@2.26:", type=("build", "link", "run"))
+    depends_on("python@3.9:3.11", when="@1.25", type=("build", "link", "run"))
     depends_on("python@3.8:3.11", when="@1.23.2:1.24", type=("build", "link", "run"))
     depends_on("python@3.8:3.10", when="@1.22:1.23.1", type=("build", "link", "run"))
     depends_on("python@:3.10", when="@1.21.2:1.21", type=("build", "link", "run"))
     depends_on("python@:3.9", when="@1.19.3:1.21.1", type=("build", "link", "run"))
     depends_on("python@:3.8", when="@1.17.3:1.19.2", type=("build", "link", "run"))
     depends_on("python@:3.7", when="@1.14.5:1.17.2", type=("build", "link", "run"))
 
-    # https://github.com/spack/spack/pull/32078
-    depends_on("py-setuptools@:63", type=("build", "run"))
-    depends_on("py-setuptools@:59", when="@:1.22.1", type=("build", "run"))
-    # Check pyproject.toml for updates to the required cython version
+    depends_on("py-cython@0.29.34:3.0", when="@1.26:", type="build")
```

Comment:
This is probably okay to leave without an upper bound; numpy doesn't contain much Cython code, and I don't expect Cython 3.1 to break anything.

Reply:
I try to match source as best I can. Honestly it's been frustrating to see numpy/scipy/etc. pinning dependencies to exact versions. I used to subscribe to semver pinning but now I don't even use upper bounds on my deps. I'm not sure what the best solution is here.

Reply:
I concur, it's frustrating and a time sink to manage. If PyPI had a better design, we wouldn't need to use upper bounds except when things are known to be broken. In Spack it's a 1 line patch to fix a break, while in PyPI it is impossible to fix. In the absence of that known breakage, I think the rule of thumb is:
- for build dependencies, use upper bounds to select versions that are known to work,
- for runtime dependencies, either do not use upper bounds or set them with the right window into the future (e.g. the N+3 rule for numpy)

Maybe we should start documenting reasons for upper bounds in pyproject.toml in a structured manner? That way packagers can know whether the bounds need to be adhered to or not. For numpy, the `setuptools` bounds were necessary due to incompatibilities, but the other build dep bounds are preemptive.

Reply:
Alright, let's change it to just 3 (who knows what will break in 4). I think it makes more sense to be flexible in the case of numpy/scipy because:

1. We happen to have a numpy/scipy dev who can review these PRs
2. It's the root of many scientific stacks and needs to be as compatible as possible

I wonder if we should do something similar for meson in py-scipy...